posts/simprocs-intro.md

```diff
@@ -0,0 +1,550 @@
+- If `P' = False` then return the simplified expression `b` and the proof `ite_cond_eq_false r` that `ite P a b = b`
+- Otherwise, let `simp` continue the search.
+
+Source code is
+```lean
+builtin_simproc ↓ [simp, seval] reduceIte (ite _ _ _) := fun e => do
+  let_expr f@ite α c i tb eb ← e | return .continue
+  let r ← simp c
+  if r.expr.isTrue then
+    let pr    := mkApp (mkApp5 (mkConst ``ite_cond_eq_true f.constLevels!) α c i tb eb) (← r.getProof)
+    return .visit { expr := tb, proof? := pr }
+  if r.expr.isFalse then
+    let pr    := mkApp (mkApp5 (mkConst ``ite_cond_eq_false f.constLevels!) α c i tb eb) (← r.getProof)
+    return .visit { expr := eb, proof? := pr }
+  return .continue
+```
+
+Intuitively, one can think of this as a "parametric" lemma that allows one to vary the right hand side depending on the value of the left hand side.
+In the case of `reduceIte`, this allows us to combine `ite_cond_eq_true` and `ite_cond_eq_false` into a single procedure that `simp` can call.
+Notice that on can use `ite_cond_eq_true` (resp `ite_cond_eq_false`) instead of `↓reduceIte`, at the cost of introducing more steps in the simplification procedure:
+
+```
+set_option trace.Meta.Tactic.simp true in
+example : ite (1≠1) (2^3 * 0 * 50 * 50) (3^2) = 3^2 := by
+  -- 6 steps: 5 rewrites and one goal discharge
+  simp only [ite_cond_eq_false, ne_eq, not_true_eq_false]
+
+set_option trace.Meta.Tactic.simp true in
+example : ite (1≠1) (2^3 * 0 * 50 * 50) (3^2) = 3^2 := by
+  -- 4 rewrites
+  simp
+```
+
+### Computation: The `reduceDvd` simproc
+
+The `reduceDvd` simproc is designed to take expressions of the form `a | b` where `a, b` are natural number literals, and simplify them to `True` or `False`.
+
+```lean
+example : 3 ∣ 9 := by
+  simp only [Nat.reduceDvd]
+
+example : ¬ 2 ∣ 49 := by
+  simp only [Nat.reduceDvd, not_false_iff]
+
+example : a ∣ a * b := by
+  simp only [Nat.reduceDvd] --fails
+```
+
+Here the metaprogram run by `Nat.reduceDvd` does the following whenever an expression of the form `a ∣ b` (where `a, b` are natural numbers) is encountered:
+- Check that `∣` is the usual natural number division.
+- Try to extract explicit (literal) values for `a` and `b`.
+- Compute `b % a`.
+- If this quantity is zero, then return `True` together with the proof `Nat.dvd_eq_true_of_mod_eq_zero a b rfl`.
+- If this quantity isn't zero, then return `False` together with the proof `Nat.dvd_eq_false_of_mod_ne_zero a b rfl`.
+
+```lean
+builtin_simproc [simp, seval] reduceDvd ((_ : Nat) ∣ _) := fun e => do
+  let_expr Dvd.dvd _ i a b ← e | return .continue
+  let some va ← fromExpr? a | return .continue
+  let some vb ← fromExpr? b | return .continue
+  if vb % va == 0 then
+    return .done { expr := mkConst ``True, proof? := mkApp3 (mkConst ``Nat.dvd_eq_true_of_mod_eq_zero) a b reflBoolTrue}
+  else
+    return .done { expr := mkConst ``False, proof? := mkApp3 (mkConst ``Nat.dvd_eq_false_of_mod_ne_zero) a b reflBoolTrue}
```

Comment:
Again I'm super-dubious about you dumping 6 lines of meta code with absolutely 0 explanation of anything. At the very least each line should be commented. Even better: all the meta code so far should be moved into a section after this one which starts "now let's take a look at the meta code which actually does this. Warning: in this section we will assume some experience with metaprogramming". Perhaps at the end of the easy section you can give a big list of current simprocs saying what they do.